### PR TITLE
Initial docs for template files

### PR DESCRIPTION
**Pull request by markbiek**
### What and why? 🤔
When I was starting with StreamBuilder, the construction of template files was very confusing to me. Now that I have a better understanding, I wanted to write up some of the things I've learned in the hopes that other beginners will find it useful.

### Testing Steps ✍️

Read through the new document to make sure it makes sense.

### You're it! 👑

@Automattic/stream-builders

### CODE REVIEW

File: docs/TemplateFiles.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
+# StreamBuilder Template Files
+
+As you've probably seen in the [Beginner's Guide](StreamBuilder-Beginners-Guide.md), streams start with a [YAML](https://yaml.org/) template file describing how all of the pieces of a stream (or streams) fit together.
+
+Larger template files may seem intimidating at first. This document is to help describe the basics of template files so you can start crafting your own.
+
+We're reference a few different types of stream classes here but we won't go into too much detail about how they work in this document. Please refer to the Beginner's Guide for some basics about different stream types.
```
Comment (sixhours) [inline docs/TemplateFiles.md:7]: `We're reference` -> `We're referencing` ?

File: docs/TemplateFiles.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	_type: Tumblr\StreamBuilder\Streams\FilteredStream
+	stream:
+		_type: Tumblr\StreamBuilder\Streams\NullStream
+	stream_filter:
+		_type: MyApplication\StreamFilters\MyFilter
+```
+
+You can see that the primary `_type` of our template is an `InjectedStream`. We now have an `injector` property which defines the `_type` of the injector that we're using along with some required properties for the Injector.
+
+First we have the injector `allocator` which determins _where_ in the stream, we'll be injecting our data. In this case, we will always insert our data at positions 0, 5, and 10.
```
Comment (sixhours) [inline docs/TemplateFiles.md]:
```suggestion
First we have the injector `allocator` which determines _where_ in the stream, we'll be injecting our data. In this case, we will always insert our data at positions 0, 5, and 10.
```

File: docs/TemplateFiles.md (hunk 1 above)
Comment (sixhours) [inline docs/TemplateFiles.md]:
```suggestion
We're referencing a few different types of stream classes here but we won't go into too much detail about how they work in this document. Please refer to the Beginner's Guide for some basics about different stream types.
```

File: docs/TemplateFiles.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+Last, we hae the injector `inner` which defines the stream class that returns the actual data we're injecting.
```
Comment (sixhours) [inline docs/TemplateFiles.md]:
```suggestion
Last, we have the injector `inner` which defines the stream class that returns the actual data we're injecting.
```

**Review by sixhours — APPROVED** [review]
This was super helpful! I made some very minor suggestions.

**Comment by markbiek** [comment]
@sixhours Thanks for catching those typos!

File: docs/TemplateFiles.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+We're reference a few different types of stream classes here but we won't go into too much detail about how they work in this document. Please refer to the Beginner's Guide for some basics about different stream types.
+
+## Your First Template File
+
+Let's start with a very basic template file.
+
+```
+_type: Tumblr\StreamBuilder\Streams\NullStream
+```
+
```
Comment (lucila) [inline docs/TemplateFiles.md:16]: I love this "hello world" example. thank you!

File: docs/TemplateFiles.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+_type: Tumblr\StreamBuilder\Streams\FilteredStream
+stream:
+	_type: Tumblr\StreamBuilder\Streams\NullStream
+stream_filter:
+	_type: MyApplication\StreamFilters\MyFilter
+```
+
+Notice how we've essentially wrapped our original `NullStream` so it's _inside_ the `FilteredStream`. 
+
+The primary `_type` property of our file is now the `FilteredStream` and our `NullStream` is inside the filter under the `stream:` property. The `FilteredStream` also has an additional `stream_filter` property which defines the class we're using to handle the actual filtering.
```
Comment (lucila) [inline docs/TemplateFiles.md:37]:
```suggestion
The primary `_type` property of our file is now the `FilteredStream` and our `NullStream` is inside the filtered stream under the `stream:` property. The `FilteredStream` also has an additional `stream_filter` property which defines the class we're using to handle the actual filtering.
```

File: docs/TemplateFiles.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+public function __construct(
+        Stream $inner,
+        StreamFilter $filter,
+        string $identity,
+        int $retry_count = null,
+        float $overfetch_ratio = null,
+        bool $skip_filters = false,
+        bool $slice_result = true
+    ) {
+```
```
Comment (lucila) [inline docs/TemplateFiles.md:61]: can you link to the "from_template" method instead? `from_template` will give you the name of the properties

**Comment by markbiek** [comment]
@lucila Thanks for the feedback. I've made those changes in #20

**Pull request merged**